File: src/dev_health_ops/external_ingest/validate.py
"""Deep per-record validation over external-ingest wire schemas (CHAOS-2691).

Single owner (master-spec CC17): created COMPLETE here in wave 1, powers
``POST /api/v1/external-ingest/validate``. CHAOS-2697's worker imports
``validate_records`` UNCHANGED in wave 4 so the endpoint and the durable
worker path can never diverge on what "valid" means.

Shape-only: does not resolve cross-batch references (``repositoryExternalId``,
``sourceExternalKey``/``targetExternalKey``, etc.) against ClickHouse — that
requires a DB round-trip the request-validation path deliberately avoids.
Kind×system-matrix / source-instance-scope enforcement (master-spec CC6) is
also out of scope here — it needs org/source context this module doesn't
have and is added by CHAOS-2695/2697's worker-side validation extension.
"""

from __future__ import annotations

from pydantic import ValidationError

from dev_health_ops.api.external_ingest.schemas import (
    RECORD_KIND_MODELS,
    RecordEnvelope,
    ValidationErrorItem,
)

#: Pydantic error "type" -> our stable per-record rejection code vocabulary
#: (master-spec CC16 lists missing_required_field/invalid_literal/unknown_kind
#: as the canonical examples; anything else collapses to invalid_field).
_MISSING_ERROR_TYPES = {"missing"}
_LITERAL_ERROR_TYPES = {"literal_error", "enum"}


def _error_code_for(pydantic_error_type: str) -> str:
    if pydantic_error_type in _MISSING_ERROR_TYPES:
        return "missing_required_field"
    if pydantic_error_type in _LITERAL_ERROR_TYPES:
        return "invalid_literal"
    return "invalid_field"


def _error_path(index: int, loc: tuple) -> str:
    base = f"records[{index}].payload"
    if not loc:
        return base
    return base + "." + ".".join(str(part) for part in loc)


def validate_records(records: list[RecordEnvelope]) -> list[ValidationErrorItem]:
    """Validate each record's payload against its kind's Pydantic model.

    Returns one ``ValidationErrorItem`` per Pydantic field error found
    (a single malformed record can produce multiple items). An unknown
    ``kind`` produces exactly one ``unknown_kind`` item and skips payload
    validation for that record (there is no model to validate against).
    """
    errors: list[ValidationErrorItem] = []
    for index, record in enumerate(records):
        model = RECORD_KIND_MODELS.get(record.kind)
        if model is None:
            errors.append(
                ValidationErrorItem(
                    index=index,
                    kind=record.kind,
                    code="unknown_kind",
                    message=f"Unknown record kind: {record.kind!r}",
                    path=f"records[{index}].kind",
                )
            )
            continue
        try:
            model.model_validate(record.payload)
        except ValidationError as exc:
            for err in exc.errors():
                errors.append(
                    ValidationErrorItem(
                        index=index,
                        kind=record.kind,
                        code=_error_code_for(err["type"]),
                        message=err["msg"],
                        path=_error_path(index, err["loc"]),
                    )
                )
    return errors


__all__ = ["validate_records"]

File: internal/api/externalingest/validate.go
package externalingest

import (
	"fmt"
	"strconv"
	"strings"

	"github.com/full-chaos/dev-health-ops/internal/api/pyjson"
	"github.com/full-chaos/dev-health-ops/internal/pythonparity"
)

// ValidationErrorItem mirrors schemas.py's ValidationErrorItem: one
// rejection reason for one record in a submitted batch.
type ValidationErrorItem struct {
	Index   int    `json:"index"`
	Kind    string `json:"kind"`
	Code    string `json:"code"`
	Message string `json:"message"`
	Path    string `json:"path,omitempty"`
}

// toPyJSON builds the ordered wire shape (schemas.py's ValidationErrorItem
// field order: index, kind, code, message, path) for policy.WriteJSON. Path
// is omitted when empty, matching the struct tag's omitempty.
func (item ValidationErrorItem) toPyJSON() *pyjson.Object {
	object := pyjson.NewObject()
	object.Set("index", item.Index)
	object.Set("kind", item.Kind)
	object.Set("code", item.Code)
	object.Set("message", item.Message)
	if item.Path != "" {
		object.Set("path", item.Path)
	}
	return object
}

// RecordInput is one record for ValidateRecords: its kind and its payload
// object, keys in input order and numbers exact.
type RecordInput struct {
	Kind    string
	Payload *pyjson.Object
}

// ValidateRecords is external_ingest/validate.py validate_records, exact in
// codes, messages and paths: an unknown kind is one unknown_kind item;
// otherwise every pydantic error of model.model_validate(payload) (python
// mode, see validateModel) becomes one item, in pydantic's order. The one
// implementation both /validate routes and the accept path use.
func ValidateRecords(records []RecordInput) []ValidationErrorItem {
	var items []ValidationErrorItem
	for index, record := range records {
		model, known := recordModels[record.Kind]
		if !known {
			items = append(items, ValidationErrorItem{
				Index: index, Kind: record.Kind, Code: "unknown_kind",
				Message: "Unknown record kind: " + pythonparity.StrRepr(record.Kind),
				Path:    fmt.Sprintf("records[%d].kind", index),
			})
			continue
		}
		payload := record.Payload
		if payload == nil {
			payload = pyjson.NewObject()
		}
		for _, err := range validateModel(model, payload) {
			items = append(items, ValidationErrorItem{
				Index: index, Kind: record.Kind, Code: errorCode(err.Type), Message: err.Msg,
				Path: errorPath(index, err.Loc),
			})
		}
	}
	return items
}

// validateRecords is ValidateRecords over parsed envelope records.
func validateRecords(records []Record) []ValidationErrorItem {
	inputs := make([]RecordInput, len(records))
	for index, record := range records {
		payload := record.ordered
		if payload == nil {
			payload = objectFromMap(record.Payload)
		}
		inputs[index] = RecordInput{Kind: record.Kind, Payload: payload}
	}
	return ValidateRecords(inputs)
}

// errorCode is validate.py _error_code_for.
func errorCode(pydanticType string) string {
	switch pydanticType {
	case "missing":
		return "missing_required_field"
	case "literal_error", "enum":
		return "invalid_literal"
	}
	return "invalid_field"
}

// errorPath is validate.py _error_path: records[i].payload, then each loc
// part as str() prints it, joined with ".".
func errorPath(index int, loc []pyjson.Value) string {
	parts := []string{fmt.Sprintf("records[%d].payload", index)}
	for _, part := range loc {
		switch typed := part.(type) {
		case string:
			parts = append(parts, typed)
		case int64:
			parts = append(parts, strconv.FormatInt(typed, 10))
		default:
			parts = append(parts, fmt.Sprint(typed))
		}
	}
	return strings.Join(parts, ".")
}
